src/main.cpp

```diff
@@ -5310,12 +5312,16 @@ uint32_t MaxLegacySigops(uint32_t nBlockTime)
     return std::numeric_limits<uint32_t>::max(); // Use accurately-counted limit
 }
 
-uint32_t ForkBits(uint32_t nTime) {
+uint32_t ForkBits(uint32_t nTime, const Consensus::Params& consensusParams) {
     uint32_t bits = 0;
     AssertLockHeld(cs_main);
     // Vote for 2 MB until the fork time to show continued support
-    if (sizeForkTime.load() < nTime && (GetBoolArg("-vote2mb", DEFAULT_2MB_VOTE) || (pblocktree->ForkBitActivated(FORK_BIT_2MB) != uint256())))
+    if (nTime < sizeForkTime.load() &&
+        nTime < consensusParams.SizeForkExpiration() &&
+        (GetBoolArg("-vote2mb", DEFAULT_2MB_VOTE) || (pblocktree->ForkBitActivated(FORK_BIT_2MB) != uint256())))
```

Comment:
You're right, we can get rid of the cs_main lock. I wrote that assert because I was planning on doing MedianTimePast, and forgot to remove it later.

Yes, it can be moved to miner.cpp.

As for reserving (and using) the bit until 2018 regardless of the fork outcome, doing so would be a departure from the BIP9 specification. I do not see how the bit would be unusable after the fork completes if the fork completes successfully. Can you elaborate?